src/awkward1/operations/convert.py

```diff
@@ -1139,6 +1139,180 @@ def getfunction(layout, depth):
     else:
         return out
 
+def to_arrow(layout):
+    import pyarrow
+
+    def recurse(layout, mask = None):
+
+        if isinstance(layout, awkward1.layout.NumpyArray):
+            numpy_arr = numpy.asarray(layout)
+            if(numpy_arr.ndim == 1):
+                return pyarrow.array(numpy_arr, mask=mask)
+            else:
+                return pyarrow.Tensor.from_numpy(numpy_arr)
+        
+        elif isinstance(layout, awkward1.layout.EmptyArray):
+            return pyarrow.array(numpy.array([], dtype=numpy.float64))
+
+        elif isinstance(layout, (awkward1.layout.IndexU8, 
+                                awkward1.layout.IndexU32, 
+                                awkward1.layout.Index8, 
+                                awkward1.layout.Index32, 
+                                awkward1.layout.Index64)):
+            return pyarrow.array(numpy.asarray(layout))
+        
+        elif isinstance(layout, awkward1.layout.ListOffsetArray32):
+
+            if (awkward1.operations.describe.parameters(layout).get("__array__") == "bytestring"):
+                arr = [awkward1.behaviors.string.ByteBehavior(
+                    layout[i]).__bytes__() for i in range(len(layout))]
+                if mask is not None:
+                    return pyarrow.array(arr, mask=numpy.array(mask, dtype=numpy.bool))
+                else:
+                    return pyarrow.array(arr)
+
+            if (awkward1.operations.describe.parameters(layout).get("__array__") == "string"):
+                arr = [awkward1.behaviors.string.CharBehavior(
+                    layout[i]).__str__() for i in range(len(layout))]
+                if mask is not None:
+                    return pyarrow.array(arr, mask=numpy.array(mask, dtype=numpy.bool))
```

Comment:
There might be a bug here which I suspect. So, in this case, the offsets are in terms of characters, right? Is this always the case whenever the `ListOffestArray` has a property `string`? And are further offsets wrapped around a higher `ListOffsetArray`? If this is the case, this should work fine, otherwise, we might need to discuss this.

Reply:
One thing that we'll need to talk about is this list comprehension: you shouldn't ever be doing a Python iteration like this (for loops or implicit for loops). The Arrow representation of byte-strings and strings matches Awkward's, so whole `content` arrays can be dropped at once. There's no reason to iterate over them, much less to convert them into Python string objects (which is very expensive).

A list of strings in both Awkward and Arrow is represented by `offsets + offsets + raw content`. In Awkward, a string is a ListArray/ListOffsetArray with a special `parameter` and in Arrow, it's a first-class type, but that's just a difference in how the metadata are expressed; the data buffers are in the same format.

There's also an issue that the primary constructors in pyarrow expect Python objects, and we often have to work around them to enter the data as buffers. The [awkward0 code](https://github.com/scikit-hep/awkward-array/blob/master/awkward/arrow.py) solved a lot of those problems; it's a good guide.

Reply:
Yeah, I fixed this issue in the recent commit. The awkward0 code was a good guide as to how to go around this issue.